com.ca.lsp.cobol/lsp-service-cobol/src/main/java/com/ca/lsp/cobol/service/OutlineService.java

```diff
@@ -0,0 +1,137 @@
+package com.ca.lsp.cobol.service;
+
+import com.broadcom.lsp.domain.cobol.databus.api.DataBusBroker;
+import com.broadcom.lsp.domain.cobol.event.api.EventObserver;
+import com.broadcom.lsp.domain.cobol.event.model.AnalysisFinishedEvent;
+import com.broadcom.lsp.domain.cobol.event.model.DataEventType;
+import com.ca.lsp.core.cobol.semantics.outline.NodeType;
+import com.ca.lsp.core.cobol.semantics.outline.OutlineNode;
+import com.ca.lsp.core.cobol.semantics.outline.OutlinePosition;
+import com.ca.lsp.core.cobol.semantics.outline.OutlineRange;
+import com.google.inject.Inject;
+import com.google.inject.assistedinject.Assisted;
+import com.google.inject.internal.cglib.core.$CollectionUtils;
+import lombok.extern.slf4j.Slf4j;
+import org.eclipse.lsp4j.*;
+import org.eclipse.lsp4j.jsonrpc.messages.Either;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class OutlineService implements EventObserver<AnalysisFinishedEvent> {
+
+  private transient DataBusBroker databus;
+  private transient MyTextDocumentService textDocumentService;
+  private String documentUri;
+  private transient CompletableFuture<MyDocumentModel> waitForDocumentModel;
+
+  @Inject
+  public OutlineService(
+      DataBusBroker databus,
+      MyTextDocumentService textDocumentService,
+      @Assisted("documentUri") String documentUri) {
+    this.databus = databus;
+    this.textDocumentService = textDocumentService;
+    this.documentUri = documentUri;
+    this.waitForDocumentModel = new CompletableFuture<>();
+  }
+
+  public List<Either<SymbolInformation, DocumentSymbol>> collectSymbols() {
+    return collectSymbols(extractDocumentModel());
+  }
+
+  static List<Either<SymbolInformation, DocumentSymbol>>collectSymbols(MyDocumentModel documentModel) {
+    if (documentModel == null || documentModel.getAnalysisResult().getOutlineTree() == null) {
```

Comment:
is getAnalysisResult() safe for NPE ?

Reply:
Good question. Looks like we never put the null as analysisResult value. Usually, in the rest of the code, no one checks for null.

Reply:
It is safe, but we have an unused constructor that may allow injecting a null AnalysisResult. it should be deleted by another PR